lib/usb/usb_efm32hg.c

```diff
@@ -0,0 +1,468 @@
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2011 Gareth McMullin <gareth@blacksphere.co.nz>
+ * Copyright (C) 2015 Kuldeep Singh Dhaka <kuldeepdhaka9@gmail.com>
+ * Copyright (C) 2018 Seb Holzapfel <schnommus@gmail.com>
+ *
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <string.h>
+#include <libopencm3/cm3/common.h>
+#include <libopencm3/efm32/memorymap.h>
+#include <libopencm3/efm32/cmu.h>
+#include <libopencm3/efm32/usb.h>
+#include <libopencm3/usb/usbd.h>
+#include "usb_private.h"
+
+/* Receive FIFO size in 32-bit words. */
+#define RX_FIFO_SIZE 256
+
+/* FIXME: EFM32HG has 6 bidirectional endpoints
+ * See detailed comment in usb_efm32lg.c */
+
+#define ENDPOINT_COUNT 4
+
+static struct _usbd_device _usbd_dev;
+
+static void efm32hg_set_address(usbd_device *usbd_dev, uint8_t addr)
+{
+	(void)usbd_dev;
+
+	USB_DCFG = (USB_DCFG & ~USB_DCFG_DEVADDR_MASK) | (addr << 4);
+}
+
+/** Initialize the USB device controller hardware of the EFM32HG. */
+static usbd_device *efm32hg_usbd_init(void)
+{
+    CMU_HFCORECLKEN0 = CMU_HFCORECLKEN0_LE | CMU_HFCORECLKEN0_USB | CMU_HFCORECLKEN0_USBC;
+
+    /* Select LFRCO as LFCCLK clock */
+    CMU_LFCLKSEL = (CMU_LFCLKSEL & ~0x30UL) | (0x1 /* LFRCO */ << 4 /* LFC */);
+
+    CMU_LFCCLKEN0 |= CMU_LFCCLKEN0_USBLE;
+
+    CMU_LFCLKSEL = (CMU_LFCLKSEL & ~CMU_LFCLKSEL_LFC_MASK) | CMU_LFCLKSEL_LFC_LFRCO;
+    CMU_LFCCLKEN0 |= CMU_LFCCLKEN0_USBLE;
+
+    // Calibrate USB based on communications
```

Comment:
Don't mix in c++ comments here.  keep them all the same style

Reply:
This was fixed by a later commit in the chain (and a good reason for me to clean the commit structure up as you suggested!)